### ISSUE TITLE
Redirect requests for replaced Whitehall attachments

### ISSUE THREAD

**Issue by floehopper** [alphagov/asset-manager#487]
In Whitehall the `AttachmentsController#show` action will redirect requests for replaced attachments . See https://github.com/alphagov/whitehall/blob/68f93f28ac1d40fedc9b49884ec30c599252cb71/app/controllers/attachments_controller.rb#L45-L47. We'll need to reproduce this behaviour for Whitehall assets served from Asset Manager.

Note that we'll need to make changes so this working on both an ongoing basis and then doing it for all the existing assets. We might want to split this into two issues.

**Comment by floehopper** [comment]
I've opened https://github.com/alphagov/asset-manager/pull/499 to add functionality to Asset Manager to cope with this. I plan to open a corresponding PR against Whitehall to make use of this new functionality.

**Comment by floehopper** [comment]
* Merged #499 
* Opened and merged #503 
* Opened alphagov/whitehall#3827

**Comment by floehopper** [comment]
https://github.com/alphagov/whitehall/pull/3827 merged

**Comment by floehopper** [comment]
I've just tested the combination of the above PRs on integration as follows:

### Before replacement

```
$ curl -s -v -H "Host: asset-manager.integration.govuk-internal.digital" "http://localhost/government/uploads/system/uploads/attachment_data/file/672667/gds-assets.pdf" >/dev/null
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET /government/uploads/system/uploads/attachment_data/file/672667/gds-assets.pdf HTTP/1.1
> User-Agent: curl/7.35.0
> Accept: */*
> Host: asset-manager.integration.govuk-internal.digital
> 
< HTTP/1.1 200 OK
* Server nginx is not blacklisted
< Server: nginx
< Date: Fri, 02 Mar 2018 17:59:16 GMT
< Content-Type: application/pdf
< Content-Length: 410572
< Connection: close
< Content-Disposition: inline; filename="gds-assets.pdf"
< Cache-Control: max-age=14400, public
< Accept-Ranges: bytes
< ETag: "5a9986b0-643cc"
< Last-Modified: Fri, 02 Mar 2018 17:15:28 GMT
< X-Frame-Options: SAMEORIGIN
< 
{ [data not shown]
* Closing connection 0
```

### After replacement

#### Original attachment

```
$ curl -s -v -H "Host: asset-manager.integration.govuk-internal.digital" "http://localhost/government/uploads/system/uploads/attachment_data/file/672667/gds-assets.pdf" >/dev/null
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET /government/uploads/system/uploads/attachment_data/file/672667/gds-assets.pdf HTTP/1.1
> User-Agent: curl/7.35.0
> Accept: */*
> Host: asset-manager.integration.govuk-internal.digital
> 
< HTTP/1.1 301 Moved Permanently
* Server nginx is not blacklisted
< Server: nginx
< Date: Fri, 02 Mar 2018 18:00:53 GMT
< Content-Type: text/html; charset=utf-8
< Transfer-Encoding: chunked
< Connection: close
< X-Frame-Options: SAMEORIGIN
< X-XSS-Protection: 1; mode=block
< X-Content-Type-Options: nosniff
< Location: http://asset-manager.integration.govuk-internal.digital/government/uploads/system/uploads/attachment_data/file/672668/2018-01-14_netLec7.pdf
< Cache-Control: max-age=14400, public
< X-Request-Id: 3cfeb830-501f-41de-afc2-02600e61e96a
< X-Runtime: 0.003998
< 
{ [data not shown]
* Closing connection 0
```

#### New attachment

```
$ curl -s -v -H "Host: asset-manager.integration.govuk-internal.digital" "http://localhost/government/uploads/system/uploads/attachment_data/file/672668/2018-01-14_netLec7.pdf" >/dev/null
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET /government/uploads/system/uploads/attachment_data/file/672668/2018-01-14_netLec7.pdf HTTP/1.1
> User-Agent: curl/7.35.0
> Accept: */*
> Host: asset-manager.integration.govuk-internal.digital
> 
< HTTP/1.1 200 OK
* Server nginx is not blacklisted
< Server: nginx
< Date: Fri, 02 Mar 2018 18:01:09 GMT
< Content-Type: application/pdf
< Content-Length: 377876
< Connection: close
< Content-Disposition: inline; filename="2018-01-14_netLec7.pdf"
< Cache-Control: max-age=14400, public
< Accept-Ranges: bytes
< ETag: "5a9990e6-5c414"
< Last-Modified: Fri, 02 Mar 2018 17:59:02 GMT
< X-Frame-Options: SAMEORIGIN
< 
{ [data not shown]
* Closing connection 0
```

#### Original thumbnail

```
$ curl -s -v -H "Host: asset-manager.integration.govuk-internal.digital" "http://localhost/government/uploads/system/uploads/attachment_data/file/672667/thumbnail_gds-assets.pdf.png" >/dev/null
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET /government/uploads/system/uploads/attachment_data/file/672667/thumbnail_gds-assets.pdf.png HTTP/1.1
> User-Agent: curl/7.35.0
> Accept: */*
> Host: asset-manager.integration.govuk-internal.digital
> 
< HTTP/1.1 301 Moved Permanently
* Server nginx is not blacklisted
< Server: nginx
< Date: Fri, 02 Mar 2018 18:04:47 GMT
< Content-Type: text/html; charset=utf-8
< Transfer-Encoding: chunked
< Connection: close
< X-Frame-Options: SAMEORIGIN
< X-XSS-Protection: 1; mode=block
< X-Content-Type-Options: nosniff
< Location: http://asset-manager.integration.govuk-internal.digital/government/uploads/system/uploads/attachment_data/file/672668/thumbnail_2018-01-14_netLec7.pdf.png
< Cache-Control: max-age=14400, public
< X-Request-Id: 51bddce4-b8a5-4f29-92a9-2ecae7eca1d0
< X-Runtime: 0.004148
< 
{ [data not shown]
* Closing connection 0
```

#### Replacement thumbnail

```
$ curl -s -v -H "Host: asset-manager.integration.govuk-internal.digital" "http://localhost/government/uploads/system/uploads/attachment_data/file/672668/thumbnail_2018-01-14_netLec7.pdf.png" >/dev/null
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET /government/uploads/system/uploads/attachment_data/file/672668/thumbnail_2018-01-14_netLec7.pdf.png HTTP/1.1
> User-Agent: curl/7.35.0
> Accept: */*
> Host: asset-manager.integration.govuk-internal.digital
> 
< HTTP/1.1 200 OK
* Server nginx is not blacklisted
< Server: nginx
< Date: Fri, 02 Mar 2018 18:05:09 GMT
< Content-Type: image/png
< Content-Length: 7157
< Connection: close
< Content-Disposition: inline; filename="thumbnail_2018-01-14_netLec7.pdf.png"
< Cache-Control: max-age=14400, public
< Accept-Ranges: bytes
< ETag: "5a9990e6-1bf5"
< Last-Modified: Fri, 02 Mar 2018 17:59:02 GMT
< X-Frame-Options: SAMEORIGIN
< 
{ [data not shown]
* Closing connection 0
```

### Asset Manager Rails console

```
$ govuk_app_console asset-manager
irb> wa = WhitehallAsset.find_by(legacy_url_path: '/government/uploads/system/uploads/attachment_data/file/672667/gds-assets.pdf')
=> #<WhitehallAsset _id: 5a9986b05381ea215e24e39a, deleted_at(deleted_at): nil, created_at: "2018-03-02T17:15:28.646Z", updated_at: "2018-03-02T17:59:34.762Z", replacement_id: BSON::ObjectId('5a9990e65d7f0978a27efe3e'), state: "uploaded", filename_history: [], uuid: "907baa05-2a83-4978-b6d3-fcb9f8cac15e", draft: false, redirect_url: nil, etag: "5a9986b0-643cc", last_modified: "2018-03-02T17:15:28.722Z", md5_hexdigest: "dc58cb59c738d3d54a431de7c5f72871", size: 410572, access_limited: [], file: "gds-assets.pdf", _type: "WhitehallAsset", legacy_url_path: "/government/uploads/system/uploads/attachment_data/file/672667/gds-assets.pdf", legacy_etag: nil, legacy_last_modified: nil>

irb> wa.public_url_path
=> "/government/uploads/system/uploads/attachment_data/file/672667/gds-assets.pdf"

irb> wa.replacement.public_url_path
=> "/government/uploads/system/uploads/attachment_data/file/672668/2018-01-14_netLec7.pdf"

irb>  wa = WhitehallAsset.find_by(legacy_url_path: '/government/uploads/system/uploads/attachment_data/file/672667/thumbnail_gds-assets.pdf.png')
=> #<WhitehallAsset _id: 5a9986b05d7f0935f44ed8d4, deleted_at(deleted_at): nil, created_at: "2018-03-02T17:15:28.532Z", updated_at: "2018-03-02T17:59:02.394Z", replacement_id: BSON::ObjectId('5a9990e65d7f09563f479621'), state: "uploaded", filename_history: [], uuid: "0bde31e3-3517-47d7-8c7f-380343a1c651", draft: false, redirect_url: nil, etag: "5a9986b0-1083", last_modified: "2018-03-02T17:15:28.575Z", md5_hexdigest: "49eb5e9c0eeb16af88553275cdf5d4b5", size: 4227, access_limited: [], file: "thumbnail_gds-assets.pdf.png", _type: "WhitehallAsset", legacy_url_path: "/government/uploads/system/uploads/attachment_data/file/672667/thumbnail_gds-assets.pdf.png", legacy_etag: nil, legacy_last_modified: nil>

irb> wa.public_url_path
=> "/government/uploads/system/uploads/attachment_data/file/672667/thumbnail_gds-assets.pdf.png"

irb> wa.replacement.public_url_path
=> "/government/uploads/system/uploads/attachment_data/file/672668/thumbnail_2018-01-14_netLec7.pdf.png"
```

I'm happy that this is all working as intended and can be deployed to staging & production.

**Comment by floehopper** [comment]
To Do:

* [ ] Test in staging/production

**Comment by chrisroos** [comment]
I'm going to check this is working staging/production now.

**Comment by chrisroos** [comment]
I've confirmed that this is working as expected in staging by:

1. Create a consultation
2. Attached a PDF
3. Observed that I could view the PDF and its thumbnail preview via the `draft-assets` host
4. Replaced the PDF with a new PDF
5. Observed that the PDF and thumbnail preview URLs in step 3 are now redirected to the asset URLs from step 4

I think this is all done so I'm closing this issue. Good work, @floehopper!
